### DISCUSSION TITLE
expression cannot fail

### DISCUSSION THREAD

**Discussion by xiayulu** [pest-parser/pest#548] [Q&A] [answered]
Hello, I am using pest parsing markdown. I write the follwing grammar:

<img width="719" alt="2021-09-16 115336" src="https://user-images.githubusercontent.com/41274826/133546910-beffa648-6511-4250-a483-8688c06ed634.png">

What's wrong with my grammar?

**Comment by CAD97** [comment, marked as answer]
PEG is greedy. Your top-level grammar is `inline = { chars | em }`.

To parse this, we first try `chars = { char* }`.

`chars` will always succeed -- it can match the empty string. Thus the `em` side of the alternation will never be tried; `chars` succeeded already.

The simple solution which probably does what you want is to instead use `chars = { char+ }`, such that `chars` must match at least one `char`, and if `char` fails to match, then `inline` will continue on to try the next alternation, `em`.
